analyzer/queries/queries.go

```diff
@@ -1,22 +1,89 @@
 package queries
 
 const (
-	LatestBlock = `
+	LatestProcessedBlock = `
     SELECT height FROM chain.processed_blocks
-      WHERE analyzer = $1
+      WHERE analyzer = $1 AND processed_time IS NOT NULL
     ORDER BY height DESC
     LIMIT 1`
 
+	FirstUnprocessedBlock = `
+    SELECT height FROM chain.processed_blocks
+      WHERE analyzer = $1 AND processed_time IS NULL
+    ORDER BY height
+    LIMIT 1`
+
+	UnlockBlockForProcessing = `
+    UPDATE chain.processed_blocks
+    SET locked_time = NULL
+    WHERE analyzer = $1 AND height = $2`
+
+	PickBlocksForProcessing = `
+    -- Locks a range of unprocessed blocks for an analyzer, handling expired locks and updating the lock status.
+    -- Parameters:
+    -- $1 = analyzer name (text)
+    -- $2 = minimum block height (integer)
+    -- $3 = maximum block height (integer, optional)
+    -- $4 = lock timeout in minutes (integer)
+    -- $5 = number of blocks to lock (integer)
+    BEGIN; -- Start a transaction, so that 'FOR UPDATE' locking persists for the whole query.
+
+    WITH
+
+    -- Find highest block that is processed, or being processed (= locked).
+    -- We'll grab the next blocks from this height on.
+    highest_done_block AS (
+      SELECT COALESCE(max(height), 0)
+      FROM chain.processed_blocks
+      WHERE analyzer = $1 AND (processed_time IS NOT NULL OR locked_time >= CURRENT_TIMESTAMP - ($4::integer * INTERVAL '1 minute'))
+    )
+
+    -- Find the lowest N($5) eligible blocks to lock.
+    blocks_to_lock AS (
+      SELECT height FROM (
+
+        -- Any unprocessed blocks with expired locks (should be few and far between).
+        SELECT height
+        FROM chain.processed_blocks
+        WHERE analyzer = $1 AND processed_time IS NULL AND (locked_time IS NULL OR locked_time < CURRENT_TIMESTAMP - ($4::integer * INTERVAL '1 minute')) AND height >= $2 AND ($3 IS NULL OR height <= $3)
+        ORDER BY height LIMIT $5
+        FOR UPDATE
+
+        UNION
+
+        -- The next 100 blocks from what is already processed or being processed.
+        SELECT height
+        FROM highest_done_block,
+          generate_series(
+            GREATEST(highest_done_block.height+1, $2), -- Don't go below $2
+            CASE                                       -- Don't go above $3, unless $3 is NULL.
+              WHEN $3 IS NOT NULL THEN LEAST(highest_done_block.height+$5, $3)
+              ELSE highest_done_block.height+$5
+            END
+          )
+        FOR UPDATE
```

Comment:
Hm, this `FOR UPDATE` is not actually locking anything here, as it is applied to the "generate_series" function. 

However, the idea here was to prevent generating the same 100 blocks, IF there are two of these queries run concurrently.

I think the simplest solution is to use an advisory lock and ensure that there is always only a single mega-query running concurrently.

---

Or do you have any other ideas @mitjat

Reply:
Ah bummer. Nice job realizing this. I agree.
An advisory lock seems to be the way to go.
I can't think of good alternatives, at least right now. I think I can come up with an ugly hacky way where the INSERT returns an extra column that signals if the inserted row was a conflict _and_ came from `generate_series`, which would then allow us to ignore those rows in Go. But a lock feels cleaner.

Reply:
Done